packages/core/src/new-api/internal/reconciliation/reconcileDependencyRules.ts

```diff
@@ -0,0 +1,44 @@
+import { difference } from "lodash";
+
+import {
+  ExecutionState,
+  ExecutionStateMap,
+  ExecutionStatus,
+} from "../../types/execution-state";
+import { Future } from "../../types/module";
+
+import { ReconciliationFutureResult } from "./types";
+import { fail } from "./utils";
+
+export function reconcileDependencyRules(
+  future: Future,
+  executionState: ExecutionState,
+  context: { executionStateMap: ExecutionStateMap }
+): ReconciliationFutureResult {
+  const previousDeps: string[] = [...executionState.dependencies];
+  const currentDeps: string[] = [...future.dependencies].map((f) => f.id);
+
+  const additionalDeps = difference(currentDeps, previousDeps);
+
+  for (const additionalDep of additionalDeps) {
+    const additionalExecutionState = context.executionStateMap[additionalDep];
+
+    if (additionalExecutionState === undefined) {
+      return fail(
+        future,
+        `A dependency from ${future.id} to the a new future ${additionalDep} has been added`
```

Comment:
```suggestion
        `A dependency from ${future.id} to ${additionalDep} has been added. The former has started executing before the latter started executing, so this change is incompatible.`
```

Reply:
I added a bit more of context to the error message here, as they weren't obvious to me, so I guess it would be less clear to users.